### DISCUSSION TITLE
[resolved] Is TerminusDB still in active development? (YES!)

### DISCUSSION THREAD

**Discussion by paulusaptus** [terminusdb/terminusdb#2133] [General]
I see that activity in this and related repositories has dropped significantly in the last year. There's a number of long-standing bugs, which do not seem to receive any attention. Hence the question: is the Terminus DB project still actively maintained?

**Comment by jprugel** [comment]
I did a little bit of digging, and it looks like it is. However, the team behind terminusdb seem to be leaning into the AI side of things, and leaving the database development to the community.

**Reply by hoijnet** [reply]
FYI: There is active development for the 12.1 branch (release candidate is now pending) with long-awaited improvements to performance, new vector search indexer and pub/sub reactivity. Since last year, high precision math, stability and many bugfixes are made continuously. Worth another look for sure.

**Comment by paulusaptus** [comment]
@jprugel can you provide any more details? It would be good to hear from anyone in the TerminusDB team on this matter, especially that it seems like the cloud infrastructure is going offline.

**Reply by hoijnet** [reply]
A commercially supported hosted version of TerminusDB is provided by us at https://dfrnt.com.

We are also working with TerminusDB on solving issues in the platform (now lately we have been fixing a double Content-Length issue that prevented full git-for-data as an example). We expect more development over time.

**Comment by adamency** [comment]
@GavinMendelGleason If you could please provide an official stance on this ?  Especially considering that given the absence of response people are now plugging their commercial clones based on your work.

**Reply by hoijnet** [reply]
This is the quoted message from the TerminusDB discord by Luke that was posted recently:

> A new exciting chapter in the development of TerminusDB opens in 2025 and into the future. 
>
> The DFRNT team is taking the helm on the continued development of the TerminusDB engine, which powers their platform. We are working to deliver a governance model that enables TerminusDB to flourish with long-term maintenance, and a strong community.
>
> @hoijnet who leads the DFRNT team has long been one of our strongest supporters; he is heavily invested into the development of TerminusDB; and continues to license the engine from us for his team's and user's needs.
>
> [DFRNT](https://dfrnt.com/) provides a complete modeller, datalog user interface, and commercial TerminusDB hosting as part of their standard business information platform. You may want to explore their offerings at [dfrnt.com](https://dfrnt.com/) .
>
> [VectorLink](https://vectorlink.ai/) offers the easiest and most scalable semantic indexing solutions available, including GPU acceleration to reduce costs and improve indexing times.  You can explore our offerings at [vectorlink.ai](https://vectorlink.ai/) .
> 
> TerminusDB will stay open source and available. A governance forum is being established and @@hoijnet, now a core Terminator, will update here as the future unfolds.
> 
> Thanks to everyone in the community for contributions and feedback so far, we've loved every part of this journey - now let the new adventures roll on under the new leadership.

You can read the full message here:
https://discord.com/channels/689805612053168129/689805612884033801/722573857638055987

We are actively working on governance for the continued development. We invite further industry and academic collaboration, and are putting in own resources as well.

**Comment by GavinMendelGleason** [comment]
TerminusDB the company had to pivot to other development work. We are still available for contract work but can't really sustain development on TerminusDB without it. There are others that are involved in development from the community such as @hoijnet

**Reply by hpvd** [reply]
Many thanks for your open words. All the best for your company and Team!!

**Reply by hoijnet** [reply]
We are actively developing TerminusDB at DFRNT and an upcoming 12.1 is soon entering release candidate stage. Stay tuned!

**Comment by hoijnet** [comment]
FYI, and with an accompanying updated welcome message on the TerminusDB Discord lately, we have a number of bug fixes and enhancements in store for 2025. 

The dfrnt.com team is taking the helm on the continued development of the TerminusDB engine, which powers our platform.

We have contracted a developer who is currently preparing to dig to enhance the engine, and thus we expect to see more releases ahead.

Read more here:
https://discord.com/channels/689805612053168129/689805612884033801/722573857638055987

So there are some good plans ahead for 2025 🎉

**Comment by hoijnet** [comment]
> I see that activity in this and related repositories has dropped significantly in the last year. There's a number of long-standing bugs, which do not seem to receive any attention. Hence the question: is the Terminus DB project still actively maintained?

Which are the most important issues for you to get resolved to take your next steps with TerminusDB?

**Reply by paulusaptus** [reply]
1. Performance is not Terminus' forte. When a DB has multiple branches with multiple commits, even mere branch listing takes a long time.
2. Docs are missing many key concepts which have to be deduced from the source code
3. Web UI is unstable and very quirky

**Reply by hoijnet** [reply]
1) Performance
As with any database engine, there are things that could be improved. The branch listings is one of them and having better support for offloading certain operations to the rust core would help reduce the efforts needed in the logic part of the query engine (driven by prolog). Prolog is fast, but not that fast. We are looking into offloading range queries to highly efficient algorithms in Rust, with Gavin's help over the coming months.

Generally, when the logic can be limited to a smaller set of operations, we recommend performance-intensive tasks to be performed using the GraphQL interface which can be up to orders of magnitudes faster than WOQL when indices can't be used. The reason is that data currently has to be materialised into the prolog part of the engine when selecting the triples from Rust, for the logic to be processed.

For many, using prolog is not practical. TerminusDB enables a practical use of many logic operations together with RDF that were previously much harder to use. It's about using the right database for the right task. 

2) Documentation
Are there any specific areas you would like to see better documentation for currently? Enhancement requests for the documentation would help us improve it over time!

3) WebUI
The built-in UI has too many bugs to be useful. We built a modeller to complement it and are currently considering what to do with the UI code that is there as it is not helpful for showcasing the strengths of the core engine. Making it more focused would probably help a lot.

As we will be working more to drive future development, now is a great time to help steer the direction of where the community wants to take the engine. Out of curiosity, what is the kind of dataset that you would like to use it for?

**Reply by jprugel** [reply]
Fwiw, I used terminusdb for a hobby project, and I found the way WOQL works to be mysterious, and I would love official documentation on it.

**Reply by hoijnet** [reply]
Thanks! Indeed, it's an area to improve and we are contemplating how to make it available the best way and how to speed up the learning process! We think about both collecting existing resources and how to form a learning path! All ideas are welcome on how to best support.

**Reply by alexander-mart** [reply]
@hoijnet It would be great to figure out how to fix stupid bugs and make things work according to the documentation :relieved: 

Such as:

- #2155
- #2156
- #2077

**Reply by alexander-mart** [reply]
@hoijnet I also believe that in order to not completely lose and slightly revive the community, we must as soon as possible bring all the starting documentation to an up-to-date form. Outdated things cause irritation and scare off newcomers who could potentially help with closing bugs and further development.

I've put together a list of some things to get we started:

- https://github.com/terminusdb/terminusdb/issues/2157
- https://github.com/terminusdb/terminusdb/issues/2158
- https://github.com/terminusdb/terminusdb-docs/issues/216
- https://github.com/terminusdb/terminusdb-docs/pull/215

**Reply by alexander-mart** [reply]
> 3\. The built-in UI has [too many bugs](https://github.com/orgs/terminusdb/discussions/2133#discussioncomment-12272855) to be useful

@hoijnet  I suggest rewriting it in [SvelteKit](https://svelte.dev/docs/kit) in a more high-quality style :)

This should be easier to maintain.

**Reply by hoijnet** [reply]
> > 3. The built-in UI has [too many bugs](https://github.com/orgs/terminusdb/discussions/2133#discussioncomment-12272855) to be useful
> 
> @hoijnet I suggest rewriting it in [SvelteKit](https://svelte.dev/docs/kit) in a more high-quality style :)
> 
> This should be easier to maintain.

I think the most important thing will be to enable easy contributions. I believe the most popular frontend library (and where we could also support to some extent) would be React. It will (probably) be easiest for many to fix issues in.

Before a new frontend is created though (it takes a lot of work), another option to consider is to prune or hide functionality in the existing one to only what works and then continue from there.

What is most important before we start is to set a few goals. All for a new SvelteKit UI in a high quality style, but it will be hard to leverage the react work that has already been done by the TerminusDB team in the past.

Not an easy one. I would lean towards React given how many developers already know it and make it easier to reuse existing code. Unless there is a possibility for community members to spend enough effort to get a SvelteKit version off the ground and be put in place to replace the existing one with enough functionality.

Documentation updates and ensuring that PRs can flow should probably be the first focus now (the DFRNT focus will be on current work on getting new builds out, and as a second priority to help documentation to be more up to date).

Thank you so much for providing the pointers. I'll see how to get them updated in the best way!

**Reply by hoijnet** [reply]
An app concept with both admin app and data app with embeddings and search indexing management are coming!

**Reply by jprugel** [reply]
> An app concept with both admin app and data app with embeddings and search indexing management are coming!

Thank you for putting so much time and effort into supporting the open source(assuming it's still open source, things can change) eco systemz this provides an immense reference for commercial grade prolog projects, which is seriously lacking due to prologs history riddled in commercial vendoring.
